### ISSUE TITLE
NCDU crashing on Android Termux

### ISSUE THREAD

**Issue by tomyummmm** [rclone/rclone#5259]
#### What is the problem you are having with rclone?
Using termux with bash, rclone ncdu crashes when i try to scroll with touchscreen or with arrow keys, producing the logs below. This causes bash to not show anything i type until i hit enter, requiring a restart or a new session. Testing with zsh also crashes, but i am able to continue typing and see it on the terminal. 

Using ncdu locally (pkg install ncdu) i can scroll and use arrow keys perfectly fine

#### What is your rclone version (output from `rclone version`)

> rclone v1.55.0-DEV
> - os/type: android
> - os/arch: arm64
> - go/version: go1.16
> - go/linking: dynamic
> - go/tags: none

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Android 10
Arm64

#### Which cloud storage system are you using? (e.g. Google Drive)

Google Team Drives

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

rclone ncdu raven:

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

> ❯ rclone ncdu raven:
> panic: runtime error: index out of range [14] with length 14
> 
> goroutine 1 [running]:
> github.com/nsf/termbox-go.Close()
>         /home/builder/.termux-build/rclone/src/.gopath/pkg/mod/github.com/nsf/termbox-go@v0.0.0-20210114135735-d04385b850e8/api.go:146 +0x458
> github.com/rclone/rclone/cmd/ncdu.(*UI).Show(0x40001f45a0, 0x0, 0x0)
>         /home/builder/.termux-build/rclone/src/cmd/ncdu/ncdu.go:842 +0x2f8
> github.com/rclone/rclone/cmd/ncdu.glob..func1.1(0x58a44a2900, 0x400004a080)
>         /home/builder/.termux-build/rclone/src/cmd/ncdu/ncdu.go:60 +0x34
> github.com/rclone/rclone/cmd.Run(0x7ffbde0000, 0x58a502d060, 0x4000b3fd40)
>         /home/builder/.termux-build/rclone/src/cmd/cmd.go:256 +0xb8
> github.com/rclone/rclone/cmd/ncdu.glob..func1(0x58a502d060, 0x4000510be0, 0x1, 0x1)
>         /home/builder/.termux-build/rclone/src/cmd/ncdu/ncdu.go:59 +0xa8
> github.com/spf13/cobra.(*Command).execute(0x58a502d060, 0x4000510bb0, 0x1, 0x1, 0x58a502d060, 0x4000510bb0)
>         /home/builder/.termux-build/rclone/src/.gopath/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:854 +0x1c8
> github.com/spf13/cobra.(*Command).ExecuteC(0x58a5024a80, 0x0, 0x4000000180, 0x4000052000)
>         /home/builder/.termux-build/rclone/src/.gopath/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:958 +0x24c
> github.com/spf13/cobra.(*Command).Execute(...)
>         /home/builder/.termux-build/rclone/src/.gopath/pkg/mod/github.com/spf13/cobra@v1.1.1/command.go:895
> github.com/rclone/rclone/cmd.Main()
>         /home/builder/.termux-build/rclone/src/cmd/cmd.go:555 +0x94
> main.main()
>         /home/builder/.termux-build/rclone/src/rclone.go:14 +0x18

Edit: screen recording uploaded here
[Screen recording](http://imgur.com/a/dGyoL9a)

**Comment by ncw** [comment]
Can you try the latest beta here: https://beta.rclone.org/v1.56.0-beta.5416.7afb4487e/

This should have android builds in `testbuilds` (when it has finished building).

This has
- a fix to ncdu
- a new version of the termbox-go library where the crash seems to to be

If that doesn't work then can you run with `--log-file rclone.log` and attach the rclone log as I'm not sure the traceback above is complete - it is certainly confusing me!

Thanks

**Comment by x0b** [comment]
Same crash (tested with https://beta.rclone.org/v1.56.0-beta.1.7afb448/testbuilds/rclone-android-21-armv8a.gz):

```
2021/04/20 07:47:48 DEBUG : Using config file from "/data/data/com.termux/files/home/rclone-internal/rclone.conf"
2021/04/20 07:47:48 DEBUG : rclone: Version "v1.56.0-beta.1.7afb448" starting with parameters ["../rclone-beta" "-vvv" "--log-file" "/data/data/com.termux/files/home/ncdu.log" "ncdu" "."]
2021/04/20 07:47:48 DEBUG : Creating backend with remote "."
2021/04/20 07:47:48 DEBUG : fs cache: renaming cache item "." to be canonical "/data/data/com.termux/files/home/rclone"
panic: runtime error: index out of range [14] with length 14

goroutine 1 [running]:
github.com/nsf/termbox-go.Close()
	github.com/nsf/termbox-go@v1.1.0/api.go:146 +0x458
github.com/rclone/rclone/cmd/ncdu.(*UI).Show(0x40001dc960, 0x0, 0x0)
	github.com/rclone/rclone/cmd/ncdu/ncdu.go:852 +0x2a0
github.com/rclone/rclone/cmd/ncdu.glob..func1.1(0x5558839600, 0x4000048040)
	github.com/rclone/rclone/cmd/ncdu/ncdu.go:60 +0x34
github.com/rclone/rclone/cmd.Run(0x7fcf620000, 0x55596c1fe0, 0x40008a7d20)
	github.com/rclone/rclone/cmd/cmd.go:267 +0xc0
github.com/rclone/rclone/cmd/ncdu.glob..func1(0x55596c1fe0, 0x4000112380, 0x1, 0x4)
	github.com/rclone/rclone/cmd/ncdu/ncdu.go:59 +0xa8
github.com/spf13/cobra.(*Command).execute(0x55596c1fe0, 0x400003d0e0, 0x4, 0x6, 0x55596c1fe0, 0x400003d0e0)
	github.com/spf13/cobra@v1.1.3/command.go:856 +0x1bc
github.com/spf13/cobra.(*Command).ExecuteC(0x55596ba2e0, 0x0, 0x5, 0x5557159248)
	github.com/spf13/cobra@v1.1.3/command.go:960 +0x244
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v1.1.3/command.go:897
github.com/rclone/rclone/cmd.Main()
	github.com/rclone/rclone/cmd/cmd.go:566 +0x90
main.main()
	github.com/rclone/rclone/rclone.go:14 +0x18
```

For me, this also hangs the termux session.

**Comment by xd003** [comment]
I am sure the logs provided above are enough. I just want to add that i too face this error on exiting ncdu process. Only difference is I don't face any termux session hang issue . I can type anything normally below that long error message once exited out of ncdu

**Comment by ncw** [comment]
I bisected this down to this commit in termbox-go: https://github.com/nsf/termbox-go/commit/b53f34a1710006fca36c1ebaabac02e9a614e9eb

Eventually I worked out what was going wrong and submitted an upstream pull request here: https://github.com/nsf/termbox-go/pull/236

Can you try this which has the fix in? It works for me so hopefully will work for you too!

[v1.56.0-beta.5417.c3d1fbdbf.fix-5259-ncdu](https://beta.rclone.org/branch/fix-5259-ncdu/v1.56.0-beta.5417.c3d1fbdbf.fix-5259-ncdu/) on branch [fix-5259-ncdu](https://github.com/rclone/rclone/tree/fix-5259-ncdu) (uploaded in 15-30 mins)

**Comment by tomyummmm** [comment]
Sorry it's my first time compiling on android, i followed the guide and installed golang and lld using apt, used the command 
> go get -u -v github,com/rclone/rclone@fix-5259-ncdu
Is this correct?

What do i do next? i backed up rclone and copied the newly built rclone into usr/bin, ran rclone but still the same error?

**Comment by x0b** [comment]
@ncw
I can no longer get it to crash, the fix forks works me. (with [c3d1fbd armv8a](https://beta.rclone.org/branch/fix-5259-ncdu/v1.56.0-beta.1.c3d1fbd.fix-5259-ncdu/testbuilds/rclone-android-21-armv8a.gz)).

**Comment by tomyummmm** [comment]
> 
> 
> @ncw
> I can no longer get it to crash, the fix forks works me. (with [c3d1fbd armv8a](https://beta.rclone.org/branch/fix-5259-ncdu/v1.56.0-beta.1.c3d1fbd.fix-5259-ncdu/testbuilds/rclone-android-21-armv8a.gz)).

Hi, could you kindly teach me how to properly compile this from the branch? I have never really used golang before nor compiled program

**Comment by tomyummmm** [comment]
@ncw 
This issue seems to be resolved, tested to be working in both bash and zsh, and doesnt cause the weird issue of bash not showing what im typing anymore. Thank you!

**Comment by ncw** [comment]
Great thank you for testing.

I'll reopen this as a reminder that we need to merge a fix from upstream.

> ...and doesnt cause the weird issue of bash not showing what im typing anymore. Thank you!

If you get that, type the command `reset` and all will be well after. You'll probably have to do it blind but it will fixup your terminal.

I've also notices that the betas for the android builds are going into the wrong place so I'll try to fix that next!

**Comment by ncw** [comment]
This fix has now been merged upstream. If this works for you I'll merge it to master :-)

[v1.56.0-beta.5424.54392fbf2.fix-5259-ncdu](https://beta.rclone.org/branch/fix-5259-ncdu/v1.56.0-beta.5424.54392fbf2.fix-5259-ncdu/) on branch [fix-5259-ncdu](https://github.com/rclone/rclone/tree/fix-5259-ncdu) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
Note the android builds should appear in the `testbuilds` subdirectory of the link above

**Comment by tomyummmm** [comment]
> This fix has now been merged upstream. If this works for you I'll merge it to master :-)
> 
> [v1.56.0-beta.5424.54392fbf2.fix-5259-ncdu](https://beta.rclone.org/branch/fix-5259-ncdu/v1.56.0-beta.5424.54392fbf2.fix-5259-ncdu/) on branch [fix-5259-ncdu](https://github.com/rclone/rclone/tree/fix-5259-ncdu) (uploaded in 15-30 mins)

I just tested with the build, everything works fine for me!

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.56 - thanks for testing
